File: SWB4/Client/OpenOffice/SWBOfficeCore/src/org/semanticwb/openoffice/ui/wizard/ContentProperties.java
/**  
* SemanticWebBuilder es una plataforma para el desarrollo de portales y aplicaciones de integración, 
* colaboración y conocimiento, que gracias al uso de tecnología semántica puede generar contextos de 
* información alrededor de algún tema de interés o bien integrar información y aplicaciones de diferentes 
* fuentes, donde a la información se le asigna un significado, de forma que pueda ser interpretada y 
* procesada por personas y/o sistemas, es una creación original del Fondo de Información y Documentación 
* para la Industria INFOTEC, cuyo registro se encuentra actualmente en trámite. 
* 
* INFOTEC pone a su disposición la herramienta SemanticWebBuilder a través de su licenciamiento abierto al público (‘open source’), 
* en virtud del cual, usted podrá usarlo en las mismas condiciones con que INFOTEC lo ha diseñado y puesto a su disposición; 
* aprender de él; distribuirlo a terceros; acceder a su código fuente y modificarlo, y combinarlo o enlazarlo con otro software, 
* todo ello de conformidad con los términos y condiciones de la LICENCIA ABIERTA AL PÚBLICO que otorga INFOTEC para la utilización 
* del SemanticWebBuilder 4.0. 
* 
* INFOTEC no otorga garantía sobre SemanticWebBuilder, de ninguna especie y naturaleza, ni implícita ni explícita, 
* siendo usted completamente responsable de la utilización que le dé y asumiendo la totalidad de los riesgos que puedan derivar 
* de la misma. 
* 
* Si usted tiene cualquier duda o comentario sobre SemanticWebBuilder, INFOTEC pone a su disposición la siguiente 
* dirección electrónica: 
*  http://www.semanticwebbuilder.org
**/ 
 
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */

/*
 * ContentProperties.java
 *
 * Created on 10/02/2009, 07:20:58 PM
 */
package org.semanticwb.openoffice.ui.wizard;

import java.awt.BorderLayout;
import java.awt.Color;
import java.util.HashMap;
import java.util.Map;
import javax.swing.JLabel;
import javax.swing.JOptionPane;
import javax.swing.JPanel;
import org.netbeans.spi.wizard.Wizard;
import org.netbeans.spi.wizard.WizardPage;
import org.netbeans.spi.wizard.WizardPanelNavResult;
import org.semanticwb.office.interfaces.PropertyInfo;
import org.semanticwb.openoffice.OfficeApplication;

/**
 *
 * @author victor.lorenzana
 */
public class ContentProperties extends WizardPage
{

    private final JLabel label = new JLabel("No se tienen propiedades para este tipo de contenido, puede continuar.");
    private final JPanel panel = new JPanel();
    public static final String CONTENT_PROPERTIES = "PROPERTIES";

    public ContentProperties()
    {
        initComponents();
        //panel.setBackground(new Color(255, 255, 255));
        panel.setLayout(new BorderLayout());
        panel.add(label, BorderLayout.NORTH);

    }

    @Override
    protected void renderingPage()
    {
        Map map = this.getWizardDataMap();
        map.put(CONTENT_PROPERTIES, null);
        if (map != null && map.get(SelectCategory.REPOSITORY_ID) != null && map.get(TitleAndDescription.NODE_TYPE) != null)
        {
            String repositoryName = map.get(SelectCategory.REPOSITORY_ID).toString();
            String type = map.get(TitleAndDescription.NODE_TYPE).toString();
            loadProperties(repositoryName, type);
        }
        super.renderingPage();
    }

    public void loadProperties(String repositoryName, String type)
    {
        panelPropertyEditor1.removeProperties();
        this.remove(this.panelPropertyEditor1);
        this.remove(panel);
        this.add(panel);
        try
        {
            PropertyInfo[] props = OfficeApplication.getOfficeDocumentProxy().getContentProperties(repositoryName, type);
            if (props.length > 0)
            {
                this.remove(panel);
                this.add(this.panelPropertyEditor1);
                HashMap<PropertyInfo, Object> properties = new HashMap<PropertyInfo, Object>();
                for (PropertyInfo info : props)
                {
                    Object defaultValue = null;
                    if (info.type.equalsIgnoreCase("string"))
                    {
                        defaultValue = "";
                    }
                    if (info.type.equalsIgnoreCase("integer"))
                    {
                        defaultValue = 0;
                    }
                    if (info.type.equalsIgnoreCase("boolean"))
                    {
                        defaultValue = false;
                    }
                    properties.put(info, defaultValue);
                }
                this.panelPropertyEditor1.loadProperties(properties);
            }
        }
        catch (Exception e)
        {
            e.printStackTrace();
        }
    }

    @Override
    public WizardPanelNavResult allowNext(String arg, Map map, Wizard wizard)
    {
        return allowFinish(arg, map, wizard);
    }

    @Override
    public WizardPanelNavResult allowFinish(String arg, Map map, Wizard wizard)
    {
        WizardPanelNavResult result = WizardPanelNavResult.PROCEED;
        Map<PropertyInfo, String> properties = this.panelPropertyEditor1.getProperties();

        for (PropertyInfo prop : properties.keySet())
        {
            String value = this.panelPropertyEditor1.getProperties().get(prop);
            if (value.isEmpty() && prop.isRequired)
            {
                JOptionPane.showMessageDialog(this, "¡Debe indicar " + prop + "!", getDescription(), JOptionPane.OK_OPTION | JOptionPane.ERROR_MESSAGE);
                panelPropertyEditor1.requestFocus();
                return WizardPanelNavResult.REMAIN_ON_PAGE;
            }
        }
        int iproperties = properties.keySet().size();
        PropertyInfo[] props = new PropertyInfo[iproperties];
        Object[] values = new Object[iproperties];
        int i = 0;
        for (PropertyInfo prop : properties.keySet())
        {
            Object value = properties.get(prop);
            props[i] = prop;
            values[i] = value;
            i++;
        }
        String repositoryName = map.get(SelectCategory.REPOSITORY_ID).toString();
        String type = map.get(TitleAndDescription.NODE_TYPE).toString();
        loadProperties(repositoryName, type);
        try
        {
            OfficeApplication.getOfficeDocumentProxy().validateContentValues(repositoryName, props, values, type);
        }
        catch (Exception e)
        {
            JOptionPane.showMessageDialog(this, e.getMessage(), getDescription(), JOptionPane.OK_OPTION | JOptionPane.ERROR_MESSAGE);
            return WizardPanelNavResult.REMAIN_ON_PAGE;
        }
        map.put(CONTENT_PROPERTIES, properties);
        return result;
    }

    /** This method is called from within the constructor to
     * initialize the form.
     * WARNING: Do NOT modify this code. The content of this method is
     * always regenerated by the Form Editor.
     */
    @SuppressWarnings("unchecked")
    // <editor-fold defaultstate="collapsed" desc="Generated Code">//GEN-BEGIN:initComponents
    private void initComponents() {

        panelPropertyEditor1 = new org.semanticwb.openoffice.components.PanelPropertyEditor();

        setLayout(new java.awt.BorderLayout());
        add(panelPropertyEditor1, java.awt.BorderLayout.CENTER);
    }// </editor-fold>//GEN-END:initComponents
    // Variables declaration - do not modify//GEN-BEGIN:variables
    private org.semanticwb.openoffice.components.PanelPropertyEditor panelPropertyEditor1;
    // End of variables declaration//GEN-END:variables

    public static String getDescription()
    {
        return "Propiedades de contenido";
    }
}

File: SWB4/Client/Microsoft Office/INFOTEC WebBuilder 4/WBOffice4/Steps/ContentProperties.cs
/**  
* SemanticWebBuilder es una plataforma para el desarrollo de portales y aplicaciones de integración, 
* colaboración y conocimiento, que gracias al uso de tecnología semántica puede generar contextos de 
* información alrededor de algún tema de interés o bien integrar información y aplicaciones de diferentes 
* fuentes, donde a la información se le asigna un significado, de forma que pueda ser interpretada y 
* procesada por personas y/o sistemas, es una creación original del Fondo de Información y Documentación 
* para la Industria INFOTEC, cuyo registro se encuentra actualmente en trámite. 
* 
* INFOTEC pone a su disposición la herramienta SemanticWebBuilder a través de su licenciamiento abierto al público (‘open source’), 
* en virtud del cual, usted podrá usarlo en las mismas condiciones con que INFOTEC lo ha diseñado y puesto a su disposición; 
* aprender de él; distribuirlo a terceros; acceder a su código fuente y modificarlo, y combinarlo o enlazarlo con otro software, 
* todo ello de conformidad con los términos y condiciones de la LICENCIA ABIERTA AL PÚBLICO que otorga INFOTEC para la utilización 
* del SemanticWebBuilder 4.0. 
* 
* INFOTEC no otorga garantía sobre SemanticWebBuilder, de ninguna especie y naturaleza, ni implícita ni explícita, 
* siendo usted completamente responsable de la utilización que le dé y asumiendo la totalidad de los riesgos que puedan derivar 
* de la misma. 
* 
* Si usted tiene cualquier duda o comentario sobre SemanticWebBuilder, INFOTEC pone a su disposición la siguiente 
* dirección electrónica: 
*  http://www.webbuilder.org.mx 
**/ 
 
﻿using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.IO;
using System.Windows.Forms;
using WBOffice4.Interfaces;
using WBOffice4.Utils;
using XmlRpcLibrary;

namespace WBOffice4.Steps
{
    public partial class ContentProperties : TSWizards.BaseInteriorStep
    {
        private OfficeDocument document;
        public static readonly String CONTENT_PROPERTIES = "CONTENT_PROPERTIES";
        public ContentProperties(OfficeDocument document)
        {
            InitializeComponent();
            this.document = document;

        }

        private void ContentProperties_ShowStep(object sender, TSWizards.ShowStepEventArgs e)
        {
            String repositoryName = this.Wizard.Data[SelectCategory.REPOSITORY_ID].ToString();
            ContentType type = ((ContentType)this.Wizard.Data[TitleAndDescription.NODE_TYPE]);
            PropertyInfo[] props = OfficeApplication.OfficeDocumentProxy.getContentProperties(repositoryName, type.id);
            this.propertyEditor1.Properties = props;
        }

        private void ContentProperties_ValidateStep(object sender, CancelEventArgs e)
        {
            FileInfo zipFile = null;
            String title = this.Wizard.Data[TitleAndDescription.TITLE].ToString();
            String description = this.Wizard.Data[TitleAndDescription.DESCRIPTION].ToString();
            String categoryID = this.Wizard.Data[SelectCategory.CATEGORY_ID].ToString();
            String repositoryName = this.Wizard.Data[SelectCategory.REPOSITORY_ID].ToString();
            RepositoryInfo rep = null;
            foreach(RepositoryInfo temp in OfficeApplication.OfficeApplicationProxy.getRepositories())
            {
                if(temp.name.Equals(repositoryName))
                {
                    rep=temp;
                }
            }
            ContentType contentType = (ContentType)this.Wizard.Data[TitleAndDescription.NODE_TYPE];
            PropertyInfo[] properties = OfficeApplication.OfficeDocumentProxy.getContentProperties(repositoryName, contentType.id);
            String[] values = this.propertyEditor1.Values;
            try
            {
                OfficeApplication.OfficeDocumentProxy.validateContentValues(repositoryName, properties, values, contentType.id);
            }
            catch (Exception ue)
            {
                OfficeApplication.WriteError(ue);
                MessageBox.Show(this, ue.Message, this.Wizard.Text, MessageBoxButtons.OK, MessageBoxIcon.Error);
                return;
            }
            this.Wizard.SetProgressBarInit(3, 1, "Preparando documento para guardar...");
            try
            {
                zipFile = document.CreateZipFile();
                this.Wizard.SetProgressBarInit(3, 2, "Guardando documento en el sitio...");
                IOfficeDocument openOfficeDocument = OfficeDocument.OfficeDocumentProxy;
                openOfficeDocument.Attachments.Add(new Attachment(zipFile, zipFile.Name));
                String name = document.FilePath.Name.Replace(document.DefaultExtension, document.PublicationExtension);
                String contentID = openOfficeDocument.save(title, description, repositoryName, categoryID, document.DocumentType.ToString().ToUpper(), contentType.id, name, properties, values);
                this.Wizard.Data[TitleAndDescription.CONTENT_ID] = contentID;
                document.SaveContentProperties(contentID, repositoryName);
                this.Wizard.SetProgressBarEnd();
                WebSiteInfo[] sites=OfficeApplication.OfficeApplicationProxy.getSites();
                if (sites != null && sites.Length > 0 && rep!=null && rep.siteInfo!=null)
                {
                    DialogResult res = MessageBox.Show(this, "¿Desea publicar el contenido en una página web?", this.Wizard.Text, MessageBoxButtons.YesNo, MessageBoxIcon.Question);                                        
                    if (res == DialogResult.Yes)
                    {
                        WebSiteInfo webSiteInfo = new WebSiteInfo();
                        webSiteInfo.id = rep.siteInfo.id;
                        webSiteInfo.title = rep.siteInfo.title;
                        document.Publish(title, description, webSiteInfo);
                    }
                }
                if (OfficeApplication.MenuListener != null)
                {
                    OfficeApplication.MenuListener.DocumentPublished();
                }
            }
            catch (Exception ue)
            {
                OfficeApplication.WriteError(ue);
                MessageBox.Show(this, "Error al tratar de publicar un documento " + ue.Message, this.Wizard.Text, MessageBoxButtons.OK, MessageBoxIcon.Error);
            }
            finally
            {
                if (zipFile != null && zipFile.Exists)
                {
                    zipFile.Delete();
                }
            }
        }
    }
}
